ldap/servers/slapd/ldapi.c

```diff
@@ -0,0 +1,409 @@
+            const char *username = slapi_entry_attr_get_ref(entries[i], CONFIG_LDAPI_AUTH_USERNAME_ATTRIBUTE);
+            uid_t user_uid = slapi_entry_attr_get_uint(entries[i], "uidNumber");
+            gid_t user_gid = slapi_entry_attr_get_uint(entries[i], "gidNumber");
+
+            if (user_uid && user_gid) {
+                /* There is a fixed uid/gid use that */
+            	struct ldapi_mapping *new_mapping = (struct ldapi_mapping *)slapi_ch_calloc(1, sizeof(struct ldapi_mapping));
+                new_mapping->uid = user_uid;
+                new_mapping->gid = user_gid;
+                new_mapping->dn = slapi_entry_attr_get_charptr(entries[i], CONFIG_LDAPI_AUTH_DN_ATTRIBUTE);
+                if (mappings == NULL) {
+                    ldapi_mappings = new_mapping; /* Set the head */
+                    mappings = new_mapping;
+                } else {
+                    mappings->next = new_mapping;
+                    mappings = mappings->next;
+                }
+            } else {
+                /* Get the uid/gid from the system */
+                struct passwd *pws;
+                pws = getpwnam(username);
+                if (pws) {
+                    struct ldapi_mapping *new_mapping = (struct ldapi_mapping *)slapi_ch_calloc(1, sizeof(struct ldapi_mapping));
+                    new_mapping->uid = pws->pw_uid;
+                    new_mapping->gid = pws->pw_gid;
+                    new_mapping->dn = slapi_entry_attr_get_charptr(entries[i], CONFIG_LDAPI_AUTH_DN_ATTRIBUTE);
+                    if (mappings == NULL) {
+                        ldapi_mappings = new_mapping; /* Set the head */
+                        mappings = new_mapping;
+                    } else {
+                        mappings->next = new_mapping;
+                        mappings = mappings->next;
+                    }
+                } else {
+                    slapi_log_err(SLAPI_LOG_WARNING, "initialize_ldapi_auth_dn_mappings",
+                            "System username (%s) in entry (%s) was not found and will be ignored.\n",
+                            username, slapi_entry_get_dn(entries[i]));
+                }
+            }
+        }
+        pthread_mutex_unlock(&dn_mapping_lock);
+    }
+    slapi_free_search_results_internal(search_pb);
+    slapi_pblock_destroy(search_pb);;
+    slapi_ch_free_string(&base_dn);
+}
+
+void
+free_ldapi_auth_dn_mappings(int32_t shutdown)
+{
+    struct ldapi_mapping *mapping = ldapi_mappings;
+    struct ldapi_mapping *next_mapping = NULL;
+
+    pthread_mutex_lock(&dn_mapping_lock);
+
+    while (mapping) {
+        next_mapping = mapping->next;
+        slapi_ch_free_string(&mapping->dn);
+        slapi_ch_free((void **)&mapping);
+        mapping = next_mapping;
+    }
+
+    pthread_mutex_unlock(&dn_mapping_lock);
+
+    if (shutdown) {
+        pthread_mutex_destroy(&dn_mapping_lock);
+    }
+}
+
+int32_t
+slapd_identify_local_user(Connection *conn)
```

Comment:
Is there a way to PR_ASSERT that the connection lock is held here?

Reply:
Hmmm, I don't think so, but this function itself, or how it is called, was not changed.  Never had a problem before...